python/dolfinx_contact/parallel_mesh_ghosting.py

```diff
@@ -0,0 +1,125 @@
+# Copyright (C) 2022 Chris N. Richardson and Sarah Roggendorf
+#
+# SPDX-License-Identifier:    MIT
+
+from dolfinx.mesh import create_mesh, meshtags
+import dolfinx
+from dolfinx.cpp.mesh import entities_to_geometry, cell_num_vertices, cell_entity_type, to_type
+import numpy as np
+
+__all__ = ["create_contact_mesh"]
+
+
+def create_contact_mesh(mesh, fmarker, dmarker, tags):
+
+    tdim = mesh.topology.dim
+    num_cell_vertices = cell_num_vertices(mesh.topology.cell_type)
+    facet_type = cell_entity_type(to_type(str(mesh.ufl_cell())), tdim - 1, 0)
+    num_facet_vertices = cell_num_vertices(facet_type)
+
+    # Get cells attached to marked facets
+    mesh.topology.create_connectivity(tdim - 1, tdim)
+    mesh.topology.create_connectivity(tdim, 0)
+    fc = mesh.topology.connectivity(tdim - 1, tdim)
+    fv = mesh.topology.connectivity(tdim - 1, 0)
+    cv = mesh.topology.connectivity(tdim, 0)
+
+    facets = np.hstack([fmarker.find(tag) for tag in tags])
+    cells_to_ghost = np.unique([fc.links(f)[0] for f in facets])
+    ncells = mesh.topology.index_map(tdim).size_local
+
+    # Convert marked facets to list of (global) vertices for each facet
+    fv_indices = [sorted(mesh.topology.index_map(0).local_to_global(fv.links(f))) for f in fmarker.indices]
+    cv_indices = [sorted(mesh.topology.index_map(0).local_to_global(cv.links(c))) for c in dmarker.indices]
+
+    # Copy facets and markers to all processes
+    global_fmarkers = sum(fv_indices, [])
+    all_indices = mesh.comm.allgather(global_fmarkers)
+    all_indices = np.array(sum(all_indices, [])).reshape(-1, num_facet_vertices)
+    all_values = np.array(sum(mesh.comm.allgather(list(fmarker.values)), []))
+
+    global_dmarkers = sum(cv_indices, [])
+    all_cell_indices = mesh.comm.allgather(global_dmarkers)
+    all_cell_indices = np.array(sum(all_cell_indices, [])).reshape(-1, num_cell_vertices)
+    all_cell_values = np.array(sum(mesh.comm.allgather(list(dmarker.values)), []))
```

Comment:
This seems really expensive.
Couldn't one use a neighborhood communicator based on if a process has any marked facets of a certain side or not?

Reply:
You are right that this is really expensive. This function is only intended as a temporary fix to make thing run in parallel. There is no point in fixing individual problems if we are planning to completely replace this in a future pull request.